pkg/grpc/server.go

```diff
@@ -86,14 +86,20 @@ func (s *server) handler(stream stream, selector cache.ResourceSelector) error {
 	var version string
 
 	for {
+		config := cache.Promise{}
+		defer func() {
+			if config.Stop != nil {
+				config.Stop()
+			}
+		}()
+
 		// make a cache request
-		var configCh <-chan api.DiscoveryResponse
 		if node != nil {
-			configCh = s.config.Listen(node, selector, version)
+			config = s.config.Listen(node, selector, version)
 		}
 
 		select {
-		case resp := <-configCh:
+		case resp := <-config.Value:
```

Comment:
Why not make config.Value a method that automatically calls Stop if required.

Reply:
It technically is since there's nothing to compute once the channel returns.